File: packages/igloo-ts/src/widgets/column.ts
import { Column as WitColumn } from 'iced:app/column@0.1.0';
import type { Length } from 'iced:app/length@0.1.0';
import type { Padding } from 'iced:app/padding@0.1.0';
import type { Horizontal } from 'iced:app/alignment@0.1.0';
import type { Pixels } from 'iced:app/shared@0.1.0';
import { Element, toElement, type IntoElement, type ElementLike } from '../element.js';

/**
 * Builder for creating Column layout widgets.
 * A Column arranges its children vertically.
 *
 * The child message type is inferred and accumulated as elements are pushed, so
 * `Column.new()` needs no annotation. Pass an explicit argument —
 * `Column.new<Msg>()` — to pin the type: each `push` is then checked against
 * `Msg` directly, so a stray message fails on the offending call.
 *
 * @example
 * ```typescript
 * const col = Column.new()
 *   .spacing(10)
 *   .push(Text.new('Item 1'))
 *   .push(Text.new('Item 2'));
 * ```
 *
 * @typeParam Msg - The accumulated message type of the column's children.
 * @typeParam Bound - Upper bound each child must satisfy; `unknown` when unpinned.
 */
export class Column<Msg = never, Bound = unknown> implements IntoElement<Msg> {
  private raw: WitColumn;

  private constructor() {
    this.raw = new WitColumn();
  }

  /** Create a new empty Column builder */
  static new(): Column<never, unknown>;
  /** Create a new empty Column builder pinned to the message type `B` */
  static new<B>(): Column<never, B>;
  static new(): Column<never, unknown> {
    return new Column();
  }

  /** Create a Column with the given elements */
  static with<const M>(elements: ElementLike<M>[]): Column<M, unknown> {
    return new Column<M, unknown>().extend(elements);
  }

  /** Add an element to the column */
  push<const M extends Bound>(element: ElementLike<M>): Column<Msg | M, Bound> {
    this.raw.push(toElement(element).inner);
    return this as unknown as Column<Msg | M, Bound>;
  }

  /** Add an element conditionally */
  pushIf<const M extends Bound>(
    condition: boolean,
    element: () => ElementLike<M>,
  ): Column<Msg | M, Bound> {
    if (condition) {
      this.raw.push(toElement(element()).inner);
    }
    return this as unknown as Column<Msg | M, Bound>;
  }

  /** Add multiple elements */
  extend<const M extends Bound>(elements: ElementLike<M>[]): Column<Msg | M, Bound> {
    for (const element of elements) {
      this.raw.push(toElement(element).inner);
    }
    return this as unknown as Column<Msg | M, Bound>;
  }

  /** Set the spacing between elements in pixels */
  spacing(spacing: Pixels): this {
    this.raw.spacing(spacing);
    return this;
  }

  /** Set the padding around the column */
  padding(padding: Padding): this {
    this.raw.padding(padding);
    return this;
  }

  /** Set the column width */
  width(width: Length): this {
    this.raw.width(width);
    return this;
  }

  /** Set the column height */
  height(height: Length): this {
    this.raw.height(height);
    return this;
  }

  /** Set the maximum width in pixels */
  maxWidth(maxWidth: Pixels): this {
    this.raw.maxWidth(maxWidth);
    return this;
  }

  /** Set horizontal alignment of children */
  alignX(align: Horizontal): this {
    this.raw.alignX(align);
    return this;
  }

  /** Enable or disable clipping of content */
  clip(clip: boolean = true): this {
    this.raw.clip(clip);
    return this;
  }

  /** Convert to Element (implements IntoElement) */
  intoElement(): Element<Msg> {
    return new Element(WitColumn.intoElement(this.raw));
  }
}

File: packages/igloo-py/src/igloo_py/widgets/column.py
"""Column layout widget builder."""

from collections.abc import Callable
from typing import TYPE_CHECKING

from wit_world.imports.column import Column as WitColumn

from ..element import Element, ElementLike, IntoElement, to_element

if TYPE_CHECKING:
    from ..types.alignment import WitHorizontal
    from ..types.length import WitLength
    from ..types.padding import WitPadding


class Column(IntoElement):
    """
    Builder for creating Column layout widgets.
    A Column arranges its children vertically.

    Example:
        # Widgets can be passed directly - no into_element() needed
        col = Column.new()
            .spacing(10)
            .push(Text.new("Item 1"))
            .push(Text.new("Item 2"))
            .push(Button.new(Text.new("Click")).on_press(lambda: msg))
    """

    def __init__(self) -> None:
        self._raw = WitColumn()

    @classmethod
    def new(cls) -> "Column":
        """Create a new empty Column builder."""
        return cls()

    @classmethod
    def with_elements(cls, elements: list[ElementLike]) -> "Column":
        """Create a Column with the given elements."""
        col = cls()
        col.extend(elements)
        return col

    def push(self, element: ElementLike) -> "Column":
        """Add an element to the column."""
        self._raw.push(to_element(element).inner)
        return self

    def push_if(self, condition: bool, element: Callable[[], ElementLike]) -> "Column":
        """Add an element conditionally."""
        if condition:
            self._raw.push(to_element(element()).inner)
        return self

    def extend(self, elements: list[ElementLike]) -> "Column":
        """Add multiple elements."""
        for element in elements:
            self._raw.push(to_element(element).inner)
        return self

    def spacing(self, spacing: float) -> "Column":
        """Set the spacing between elements in pixels."""
        self._raw.spacing(spacing)
        return self

    def padding(self, padding: "WitPadding") -> "Column":
        """Set the padding around the column."""
        self._raw.padding(padding)
        return self

    def width(self, width: "WitLength") -> "Column":
        """Set the column width."""
        self._raw.width(width)
        return self

    def height(self, height: "WitLength") -> "Column":
        """Set the column height."""
        self._raw.height(height)
        return self

    def max_width(self, max_width: float) -> "Column":
        """Set the maximum width in pixels."""
        self._raw.max_width(max_width)
        return self

    def align_x(self, align: "WitHorizontal") -> "Column":
        """Set horizontal alignment of children."""
        self._raw.align_x(align)
        return self

    def clip(self, clip: bool = True) -> "Column":
        """Enable or disable clipping of content."""
        self._raw.clip(clip)
        return self

    def into_element(self) -> Element:
        """Convert to Element (implements IntoElement)."""
        return Element(WitColumn.into_element(self._raw))
